File: src/ladym/layers/procedural.py
"""L3 — Procedural memory.

Stores reusable playbooks ``preconditions → steps → expected_outcome`` plus verified code
snippets. CoALA frames procedural memory as the agent's "skill library"; Voyager is the
canonical example of agents writing new skills here.
"""

from __future__ import annotations

from ..schema import Layer, Memory, MemoryType
from ..storage.embeddings import EmbeddingProvider
from ..storage.store import SQLiteStore
from .semantic import content_hash


def _playbook_content(name: str, steps: list[str]) -> str:
    """Canonical content string for a playbook — single source of truth.

    Used by both ``put_playbook`` (when writing) and ``proceduralize`` (when computing a
    candidate hash for dedup) so the two never drift.
    """
    return name + "\n" + "\n".join(f"{i+1}. {s}" for i, s in enumerate(steps))


class ProceduralMemory:
    def __init__(self, store: SQLiteStore, embedder: EmbeddingProvider,
                 workspace: str = "default"):
        self.store = store
        self.embedder = embedder
        self.workspace = workspace

    def put_playbook(self, name: str, steps: list[str], *,
                     preconditions: list[str] | None = None,
                     expected_outcome: str = "",
                     tags: list[str] | None = None) -> Memory:
        body = {
            "name": name,
            "preconditions": preconditions or [],
            "steps": steps,
            "expected_outcome": expected_outcome,
        }
        content = _playbook_content(name, steps)
        mem = Memory(
            layer=Layer.PROCEDURAL,
            type=MemoryType.PLAYBOOK,
            content=content,
            summary=name,
            tags=(tags or []) + ["playbook"],
            metadata=body,
            source="proceduralize",
            workspace=self.workspace,
            content_hash=content_hash(content),
        )
        self.store.put_memory(mem, vector=self.embedder.embed(content))
        return mem

    def put_snippet(self, title: str, code: str, *, language: str = "python",
                    tags: list[str] | None = None) -> Memory:
        mem = Memory(
            layer=Layer.PROCEDURAL,
            type=MemoryType.SNIPPET,
            content=f"{title}\n```{language}\n{code}\n```",
            summary=title,
            tags=(tags or []) + ["snippet", language],
            metadata={"language": language, "code": code, "title": title},
            workspace=self.workspace,
        )
        self.store.put_memory(mem, vector=self.embedder.embed(mem.content))
        return mem

File: layers/procedural.go
package layers

import (
	"fmt"
	"strings"

	"github.com/ProjAnvil/LadyM/schema"
	"github.com/ProjAnvil/LadyM/storage"
)

// ProceduralMemory is L3 — reusable playbooks and verified snippets.
type ProceduralMemory struct {
	Store     *storage.SQLiteStore
	Embedder  storage.EmbeddingProvider
	Workspace string
}

// NewProceduralMemory builds a ProceduralMemory.
func NewProceduralMemory(store *storage.SQLiteStore, embedder storage.EmbeddingProvider, workspace string) *ProceduralMemory {
	if workspace == "" {
		workspace = "default"
	}
	return &ProceduralMemory{Store: store, Embedder: embedder, Workspace: workspace}
}

// PlaybookContent is the canonical content string for a playbook — the single
// source of truth shared by PutPlaybook and proceduralize. Matches Python
// “name + "\n" + "\n".join(steps)“: empty steps keeps the trailing newline;
// non-empty steps end without one.
func PlaybookContent(name string, steps []string) string {
	var sb strings.Builder
	sb.WriteString(name)
	sb.WriteString("\n")
	for i, s := range steps {
		if i > 0 {
			sb.WriteString("\n")
		}
		fmt.Fprintf(&sb, "%d. %s", i+1, s)
	}
	return sb.String()
}

// PutPlaybook writes a playbook.
func (p *ProceduralMemory) PutPlaybook(name string, steps []string, preconditions []string, expectedOutcome string, tags []string) (*schema.Memory, error) {
	if preconditions == nil {
		preconditions = []string{}
	}
	if tags == nil {
		tags = []string{}
	}
	body := map[string]any{
		"name":             name,
		"preconditions":    preconditions,
		"steps":            steps,
		"expected_outcome": expectedOutcome,
	}
	content := PlaybookContent(name, steps)
	m := schema.NewMemory(schema.LayerProcedural, schema.TypePlaybook)
	m.Content = content
	m.Summary = name
	m.Tags = append(append([]string{}, tags...), "playbook")
	m.Metadata = body
	m.Source = "proceduralize"
	m.Workspace = p.Workspace
	m.ContentHash = schema.ContentHash(content)

	vec, err := p.Embedder.Embed(content)
	if err != nil {
		return nil, err
	}
	if err := p.Store.PutMemory(m, vec); err != nil {
		return nil, err
	}
	return m, nil
}

// PutSnippet writes a verified code snippet.
func (p *ProceduralMemory) PutSnippet(title, code, language string, tags []string) (*schema.Memory, error) {
	if language == "" {
		language = "python"
	}
	if tags == nil {
		tags = []string{}
	}
	content := fmt.Sprintf("%s\n```%s\n%s\n```", title, language, code)
	m := schema.NewMemory(schema.LayerProcedural, schema.TypeSnippet)
	m.Content = content
	m.Summary = title
	m.Tags = append(append([]string{}, tags...), "snippet", language)
	m.Metadata = map[string]any{"language": language, "code": code, "title": title}
	m.Workspace = p.Workspace

	vec, err := p.Embedder.Embed(content)
	if err != nil {
		return nil, err
	}
	if err := p.Store.PutMemory(m, vec); err != nil {
		return nil, err
	}
	return m, nil
}
